hail/src/main/scala/is/hail/expr/types/physical/PArray.scala

```diff
@@ -1,43 +1,38 @@
 package is.hail.expr.types.physical
 
-import is.hail.annotations.{UnsafeUtils, _}
-import is.hail.asm4s._
-import is.hail.check.Gen
+import is.hail.annotations.CodeOrdering
 import is.hail.expr.ir.EmitMethodBuilder
 import is.hail.expr.types.virtual.TArray
-import is.hail.utils._
-import scala.reflect.{ClassTag, _}
+object PArray {
+  def apply(elementType: PType, required: Boolean = false) = new PCanonicalArray(elementType, required)
+}
 
-final case class PArray(elementType: PType, override val required: Boolean = false) extends PContainer with PStreamable {
+abstract class PArray extends PContainer with PStreamable {
   lazy val virtualType: TArray = TArray(elementType.virtualType, required)
 
-  val elementByteSize: Long = UnsafeUtils.arrayElementSize(elementType)
+  def copy(required: Boolean): PArray
+
+  def copy(elementType: PType): PArray
+
+  def _asIdent = s"array_of_${elementType.asIdent}"
+  def _toPretty = s"Array[$elementType]"
 
-  val contentsAlignment: Long = elementType.alignment.max(4)
+  def codeOrdering(mb: EmitMethodBuilder, other: PType): CodeOrdering = {
```

Comment:
I left codeOrdering on PArray, because I am not certain whether each implementation of PArray is to have its own.

Reply:
Take a look at CodeOrdering.iterableOrdering -- it happily uses PArray interfaces. It's possible we'll want to build specialized orderings sometimes (if a PArray implementation can, for instance, compare bytes), but in general this is the right design.